### PR TITLE
Implement Persistent Actor

### PR DESCRIPTION
**Pull request by simon-mo**
This PR adds a new API for actor creation: `_remote(..., is_persistent=True`. This will make sure on a normal driver exit, the persistent actor will be kept alive.

### CODE REVIEW

File: src/ray/raylet/worker.h [hunk 1]
```diff
@@ -100,6 +102,8 @@ class Worker {
   rpc::ClientCallManager &client_call_manager_;
   /// The rpc client to send tasks to this worker.
   std::unique_ptr<rpc::WorkerTaskClient> rpc_client_;
+  /// Whether the worker is persistent.
+  bool is_persistent_;
```
Comment (zhijunfu) [inline src/ray/raylet/worker.h]: this is not initialized?

**Review by zhijunfu — CHANGES_REQUESTED** [review]
seems to have a bug on variable initialization, others looks good. thanks.

File: src/ray/raylet/worker.h [hunk 2]
```diff
@@ -100,6 +102,8 @@ class Worker {
   rpc::ClientCallManager &client_call_manager_;
   /// The rpc client to send tasks to this worker.
   std::unique_ptr<rpc::WorkerTaskClient> rpc_client_;
+  /// Whether the worker is persistent.
+  bool is_detached_;
```
Comment (pcmoritz) [inline src/ray/raylet/worker.h]: Let's call this `is_detached_actor_` and say in the documentation that this is only relevant for workers that are actors :)

File: src/ray/raylet/node_manager.cc [hunk 3]
```diff
@@ -275,13 +275,15 @@ void NodeManager::HandleJobTableUpdate(const JobID &id,
...
-        // Then kill the worker process.
-        KillWorker(worker);
+        if (!worker->IsDetachedActor()) {
+          // Clean up any open ray.wait calls that the worker made.
+          task_dependency_manager_.UnsubscribeWaitDependencies(worker->WorkerId());
+          // Mark the worker as dead so further messages from it are ignored
+          // (except DisconnectClient).
+          worker->MarkDead();
+          KillWorker(worker);
+          // Then kill the worker process.
```
Comment (pcmoritz) [inline src/ray/raylet/node_manager.cc]: you should move this up one line again

File: python/ray/actor.py [hunk 4]
```diff
@@ -339,6 +343,22 @@ def _remote(self,
 
         meta = self.__ray_metadata__
 
+        if detached and name is None:
+            raise Exception("Detached actors must have associated name. "
```
Comment (pcmoritz) [inline python/ray/actor.py]: How about `"Detached actors must be named. Please use..."`?

File: python/ray/actor.py [hunk 5]
```diff
@@ -323,6 +325,8 @@ def _remote(self,
                 this actor when creating objects.
             resources: The custom resources required by the actor creation
                 task.
+            name: The globally unique name for the actor.
+            detached: Whether the actor should be kept alive after driver exits
```
Comment (pcmoritz) [inline python/ray/actor.py]: `.` at the end

File: python/ray/includes/common.pxd [hunk 6]
```diff
@@ -206,7 +206,8 @@ cdef extern from "ray/core_worker/common.h" nogil:
             uint64_t max_reconstructions, c_bool is_direct_call,
             const unordered_map[c_string, double] &resources,
             const unordered_map[c_string, double] &placement_resources,
-            const c_vector[c_string] &dynamic_worker_options)
+            const c_vector[c_string] &dynamic_worker_options,
+            c_bool is_persistent)
```
Comment (pcmoritz) [inline python/ray/includes/common.pxd]: `is_detached`?

File: python/ray/_raylet.pyx [hunk 7]
```diff
@@ -900,28 +901,32 @@ cdef class CoreWorker:
                      args,
                      uint64_t max_reconstructions,
                      resources,
-                     placement_resources):
+                     placement_resources,
+                     is_persistent):
```
Comment (pcmoritz) [inline python/ray/_raylet.pyx]: `is_detached`?

Comment (pcmoritz) [reply python/ray/_raylet.pyx]: You can actually say `c_bool is_detached` here and get rid of the `c_is_persistent` variable below.

**Comment by pcmoritz** [comment]
This looks great, I left a few minor comments and can be merged after they are fixed :)

File: src/ray/raylet/worker.h (hunk 1 above)
Comment (simon-mo) [reply src/ray/raylet/worker.h]: Resolved.

**Review by pcmoritz — APPROVED** [review]

**Review by zhijunfu — APPROVED** [review]

**Comment by raulchen** [comment]
This feature is very useful. However, this API might have problems in some cases. 

Case 1:
```
# Driver code:
counter = Counter._remote(name="CounterActor", detached=True)
# Driver can't exit immediately here, because the actor creation task might not have been executed yet and could be cleaned up.
# Thus, we have to call a dummy method  to wait until the actor get created.
ray.get(counter.foo.remote())
```

Case 2:
```
counter = Counter._remote(name="CounterActor", detached=True)
ray.get(counter.foo.remote())
x = func.remote() # x is the return of a normal task.
counter.bar.remote(x) # `func` could get cleaned up as soon as driver exits. Then `counter.bar` will never run.
```

Thus, instead of making actors detached, I think it makes more sense simply **allow drivers to exit without cleanup**. Also, we need to expose an API to allow users to manually clean up the job afterwards. 

We've already implemented this in Ant's internal version. If you think this makes sense, we're glad to upstream this feature.

**Comment by simon-mo** [comment]
@raulchen Detached actors are suppose to be used by libraries with extreme care. However, I agree that allowing drivers to exit without cleanup with also a valid option and I would be happy to see if it can be upstreamed. What would it look like? Would it be some parameters in `ray.init`? 

@stephanie-wang Would there be any object ownership concerns with a driver just exists without cleanup?

**Comment by stephanie-wang** [comment]
We have control over the driver code, so why not just implement the driver so that it waits for detached actors/tasks to finish before exiting? Or, just use the detached flag for these tasks/actors to make sure that they don't get cleaned up while queued. I'm concerned about requiring applications to manually clean up their state; realistically applications are not going to do this correctly.

In the object ownership design, it should be okay, but it makes things much easier if we know at task/actor creation time that the task/actor is going to be "detached" (as @simon-mo did here for actors). Also, the first version would not have the ability to restart such tasks/actors.

**Comment by raulchen** [comment]
> We have control over the driver code, so why not just implement the driver so that it waits for detached actors/tasks to finish before exiting? 

@stephanie-wang The problem is that not only the driver, actors could also submit normal tasks at any time. If the driver exits while the normal tasks are still being queued, those tasks will get lost. 

> Or, just use the detached flag for these tasks/actors to make sure that they don't get cleaned up while queued.

As mentioned above, sometimes it's uncertain when a normal task will be executed. That means, basically we need to add the "detached flag" to every normal task, which is inconvenient. Also, this will add extra complexities to scheduling queue code. 

@simon-mo Our current API for exiting driver without cleanup is `ray.shutdown(cleanup=False)`. We don't currently expose the API for manually killing a job yet, because all applications that need this feature are long running. So we simply shut down the cluster when the job is not needed. Does this also apply to ray serve? In the future, maybe we can allow users to kill their long running job by clicking a button on an admin web portal.

**Pull request merged**